sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -152,26 +153,54 @@ class GpuSingleDirectoryDataWriter(
   override def write(batch: ColumnarBatch): Unit = {
     var needToCloseBatch = true
     try {
-      // TODO: This does not handle batch splitting to make sure maxRecordsPerFile is not exceeded.
-      if (description.maxRecordsPerFile > 0 && recordsInFile >= description.maxRecordsPerFile) {
-        fileCounter += 1
-        assert(fileCounter < MAX_FILE_COUNTER,
-          s"File counter $fileCounter is beyond max value $MAX_FILE_COUNTER")
+      withResource(GpuColumnVector.from(batch)) {table => 
+        val numRows = table.getRowCount()
+        val maxRecordsPerFile = description.maxRecordsPerFile.toInt
+        val fillOldFile = recordsInFile < maxRecordsPerFile
+        val splitIndexes = (maxRecordsPerFile > 0, fillOldFile) match {
+          case (false, _) => IndexedSeq.empty
+          case (true, false) =>
+            (1 until ((numRows + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
+              .map(i => i * maxRecordsPerFile)
+          case (true, true) => {
+            val filledUp = maxRecordsPerFile - recordsInFile.toInt
+            val remain = numRows - filledUp
+            (0 until ((remain + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
+              .map(i => filledUp + i * maxRecordsPerFile)
+          }
+        }
 
-        newOutputWriter()
+        val dataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType()).toArray
+
+        var needNewWriter = !fillOldFile
+
+        withResource(table.contiguousSplit(splitIndexes: _*)) {tabs =>
+          tabs.foreach(b => {
+            if (needNewWriter) {
+              fileCounter += 1
+              assert(fileCounter < MAX_FILE_COUNTER,
+                s"File counter $fileCounter is beyond max value $MAX_FILE_COUNTER")
+              newOutputWriter()
+              recordsInFile = 0
+            }
+            withResource(b.getTable()) {tab =>
+              val bc = GpuColumnVector.from(tab, dataTypes)
+              statsTrackers.foreach(_.newBatch(bc))
+              recordsInFile += b.getRowCount()
+              currentWriter.write(bc, statsTrackers)
+              needNewWriter = true
+            }
+          })
+        }
+        batch.close()
```

Comment:
It's weird to be setting a variable to ensure we close the batch by default but then only set it to false at the end of the block after we close it.  I think we can just ditch `needToCloseBatch` and the `finally` block and simply use `withResource(batch)` to ensure it is always closed.

Reply:
Updated